In this project ,i am going to create a model that will be able to identify handwritten numbers from an image.<br>
For this purpose i will be using MNIST datase which contains 60000 grescale images of handwritten number of size 28x28 and the training set contain 10000 images.

The MNIST dataset is one of the most common datasets used for image classification and accessible from many different sources. In fact, even Tensorflow and Keras allow us to import and download the MNIST dataset directly from their API<br>
code,<br>
import tensorflow as tf<br>
(xtrain,xtest),(ytrain,ytest)=tf.keras.datasets.mnist.load_data()

# workflow
1. import mnist data set from tensorflow api
2. prepare data for modelling
3. create model
4. test model
5. tune model
6. save and reload model


# 1.import dataset

In [1]:
import tensorflow as tf
import pandas as pd


In [2]:
train_csv=pd.read_csv("dataset/mnist_train.csv")
test_csv=pd.read_csv("dataset/mnist_test.csv")
print("train data:",train_csv.shape)
print("test data:",test_csv.shape)
print(train_csv.head())
print(test_csv.head())

train data: (60000, 785)
test data: (10000, 785)
   label  1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  ...  28x19  28x20  \
0      5    0    0    0    0    0    0    0    0    0  ...      0      0   
1      0    0    0    0    0    0    0    0    0    0  ...      0      0   
2      4    0    0    0    0    0    0    0    0    0  ...      0      0   
3      1    0    0    0    0    0    0    0    0    0  ...      0      0   
4      9    0    0    0    0    0    0    0    0    0  ...      0      0   

   28x21  28x22  28x23  28x24  28x25  28x26  28x27  28x28  
0      0      0      0      0      0      0      0      0  
1      0      0      0      0      0      0      0      0  
2      0      0      0      0      0      0      0      0  
3      0      0      0      0      0      0      0      0  
4      0      0      0      0      0      0      0      0  

[5 rows x 785 columns]
   label  1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  ...  28x19  28x20  \
0      7    0    0    0    0    0 

In [3]:
#splitting data int features and labels
xtrain=train_csv.drop("label",axis=1)
ytrain=train_csv.label
xtest=test_csv.drop("label",axis=1)
ytest=test_csv.label
print("train split:",xtrain.shape,ytrain.shape)
print("test split:",ytest.shape,ytest.shape)

train split: (60000, 784) (60000,)
test split: (10000,) (10000,)


since our data is in csv file ,we have to convert it int an array using np.array() function
<br>the cnn convolution layer  accepts input in the format(number_of_sample,width,height,rgb_channel) so we have to reshape our array in this format

In [4]:
import numpy as np
xtrain=np.array(xtrain).reshape(xtrain.shape[0],28,28,1)
ytrain=np.array(ytrain)
xtest=np.array(xtest).reshape(xtest.shape[0],28,28,1)
ytest=np.array(ytest)
print(xtest.shape)
print(ytest.shape)
print(xtrain.shape)
print(ytrain.shape)

(10000, 28, 28, 1)
(10000,)
(60000, 28, 28, 1)
(60000,)


This process is known as "normalization" or "transformation" and is part of your feature engineering. Your problem applies to all machine learning algorithms, not just neural networks.
<br>
We usually prefer [0,1], because they are easier to deal with<br>
 there are 784 inputs (each pixel 28x28) and each value is between 0-255 (single grayscale). However, this would produce math range errors with the sigmoid function because too larger negative values would be produced on later layers. So, all value are divided by 255 to get decimal values between 0-1.

In [5]:
# Making sure that the values are float so that we can get decimal points after division
xtrain = xtrain.astype('float32')
xtest = xtest.astype('float32')
# Normalizing the RGB codes by dividing it to the max RGB value.
xtrain=xtrain/255
xtest=xtest/255

In [6]:

#steps for hypertuning the model

# from time import time
# t1=time()
# np.random.seed(42)
# from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
# from sklearn.model_selection import GridSearchCV
# from tensorflow.keras.layers import Conv2D,Dropout,Flatten,MaxPooling2D,Dense 
# from tensorflow.keras.models import Sequential

# def create_model(units,dropout,optimizer,filters):
#     model=Sequential()
#     model.add(Conv2D(filters=filters,kernel_size=(3,3),input_shape=(28,28,1)))
#     #28:number of filters
#     #kernel_size or filter matrix size
#     model.add(MaxPooling2D(pool_size=(2,2)))
#     model.add(Flatten())
#     model.add(Dense(units=units,activation="relu",))
#     model.add(Dropout(dropout))
#     model.add(Dense(units=10,activation="softmax"))
#     model.compile(optimizer=optimizer, 
#                   loss='sparse_categorical_crossentropy', 
#                   metrics=['accuracy'])
#     #model.fit(x=xtrain,y=ytrain, epochs=10)
#     return model
# param={'filters':[28,32,64],
#        'units':[128,256],
#        'dropout':[0.1,0.2,0.4],
#        'optimizer':['adam','rmsprop'],
#        'nb_epoch':[10]
#     }
# model=KerasClassifier(build_fn=create_model)
# cv=GridSearchCV(estimator=model,param_grid=param,scoring="accuracy",cv=10)

# cv.fit(xtrain,ytrain)
# t1=time()
# print("best parameters=",cv.best_params_)
# print("best score=",cv.best_score_)
# print(f"modelling take {(t2-t1)*0.000278} hrs")

In [30]:
xtrain[0]

array([[[0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ]],

       [[0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        

In [14]:
#final tuned model
np.random.seed(42)
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV
from tensorflow.keras.layers import Conv2D,Dropout,Flatten,MaxPooling2D,Dense 
from tensorflow.keras.models import Sequential

model=Sequential()
model.add(Conv2D(filters=64,kernel_size=(3,3),input_shape=(28,28,1)))
#28:number of filters
#kernel_size or filter matrix size
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Flatten())
model.add(Dense(units=256,activation="relu",))
model.add(Dropout(0.2))
model.add(Dense(units=10,activation="softmax"))
model.compile(optimizer="adam", 
              loss='sparse_categorical_crossentropy', 
              metrics=['accuracy'])
model.fit(x=xtrain,y=ytrain, epochs=25)

Train on 60000 samples
Epoch 1/25
60000/60000 [==============================] - 104s 2ms/sample - loss: 0.1650 - accuracy: 0.9507
Epoch 2/25
60000/60000 [==============================] - 103s 2ms/sample - loss: 0.0640 - accuracy: 0.9801
Epoch 3/25
60000/60000 [==============================] - 103s 2ms/sample - loss: 0.0453 - accuracy: 0.9851
Epoch 4/25
60000/60000 [==============================] - 103s 2ms/sample - loss: 0.0317 - accuracy: 0.9895
Epoch 5/25
60000/60000 [==============================] - 103s 2ms/sample - loss: 0.0264 - accuracy: 0.9911
Epoch 6/25
60000/60000 [==============================] - 102s 2ms/sample - loss: 0.0262 - accuracy: 0.9914
Epoch 7/25
60000/60000 [==============================] - 102s 2ms/sample - loss: 0.0206 - accuracy: 0.9931
Epoch 8/25
60000/60000 [==============================] - 101s 2ms/sample - loss: 0.0167 - accuracy: 0.9949
Epoch 9/25
60000/60000 [==============================] - 103s 2ms/sample - loss: 0.0188 - accuracy: 0.9940 - los

In [15]:

model.evaluate(xtest,ytest)

10000/10000 [==============================] - 4s 447us/sample - loss: 0.1379 - accuracy: 0.9836


[0.13788492248731937, 0.9836]

In [16]:
# filters=28,units=128 [loss=0.07418991162637249,accuracy= 0.9786]
#filters=64 units=256  [loss=0.05824375776508823, accuracy=0.9834]

#saving and reloading the save model

#serialize model to json file and save weights
from tensorflow.keras.models import model_from_json
model_json=model.to_json()
with open("model.json","w") as json_file:
    json_file.write(model_json)
model.save_weights("model.h5")
print("model saved")



model saved
model loaded
10000/10000 [==============================] - 4s 438us/sample - loss: 0.1379 - accuracy: 0.9836
[0.13788492248731937, 0.9836]


In [8]:
from tensorflow.keras.models import model_from_json
#load model
json_file=open('model.json','r')
loaded_model_json=json_file.read()
json_file.close()
loaded_model=model_from_json(loaded_model_json)
loaded_model.load_weights("model.h5")
print("model loaded")

loaded_model.compile(optimizer="adam",loss='sparse_categorical_crossentropy',metrics=['accuracy'])

#score=loaded_model.evaluate(xtest,ytest)
#print(score)

model loaded


In [13]:
# model.save(filepath="D:\codes\ml_and_data_science\mnist\models")
# model.save('trainedmodel.h5') 


# new_model = tf.keras.models.load_model('D:/codes/ml_and_data_science/cats_and_dog/my_model_final.h5')
# new_model.evaluate(xtest,ytest)

Instructions for updating:
If using Keras pass *_constraint arguments to layers.
INFO:tensorflow:Assets written to: D:\codes\ml_and_data_science\mnist\models\assets


OSError: SavedModel file does not exist at: D:/codes/ml_and_data_science/cats_and_dog/my_model_final.h5/{saved_model.pbtxt|saved_model.pb}

In [68]:
from tensorflow.keras.preprocessing import image
import numpy as np
def predict_num(path):
    
    try:
        test_image=image.load_img(path,target_size=(28,28),color_mode="grayscale")
        test_image=image.img_to_array(test_image)
        test_image=np.expand_dims(test_image,axis=0)
        
        #test_image=test_image.astype("float32")
        #test_image=test_image/255
        print(test_image)
        print(test_image.shape)
        #train_generator.class_indices
        result=loaded_model.predict(test_image)
        print(result.argmax())   
        return result
              
    except:
         print("invalid address")
          

In [88]:
from PIL import Image, ImageFilter


def imageprepare(argv):
    """
    This function returns the pixel values.
    The imput is a png file location.
    """
    im = Image.open(argv).convert('L')
    width = float(im.size[0])
    height = float(im.size[1])
    newImage = Image.new('L', (28, 28), (0))  # creates white canvas of 28x28 pixels

    if width > height:  # check which dimension is bigger
        # Width is bigger. Width becomes 20 pixels.
        nheight = int(round((20.0 / width * height), 0))  # resize height according to ratio width
        if (nheight == 0):  # rare case but minimum is 1 pixel
            nheight = 1
            # resize and sharpen
        img = im.resize((20, nheight), Image.ANTIALIAS).filter(ImageFilter.SHARPEN)
        wtop = int(round(((28 - nheight) / 2), 0))  # calculate horizontal position
        newImage.paste(img, (4, wtop))  # paste resized image on white canvas
    else:
        # Height is bigger. Heigth becomes 20 pixels.
        nwidth = int(round((20.0 / height * width), 0))  # resize width according to ratio height
        if (nwidth == 0):  # rare case but minimum is 1 pixel
            nwidth = 1
            # resize and sharpen
        img = im.resize((nwidth, 20), Image.ANTIALIAS).filter(ImageFilter.SHARPEN)
        wleft = int(round(((28 - nwidth) / 2), 0))  # caculate vertical pozition
        newImage.paste(img, (wleft, 4))  # paste resized image on white canvas

    newImage.save("temp.png")

    tv = list(newImage.getdata())  # get pixel values

    # normalize pixels to 0 and 1. 0 is pure white, 1 is pure black.
    tva = [(255 - x) * 1.0 / 255.0 for x in tv]
    print(tva)
    return tva

imageprepare("temp.png")

result=predict_num("temp.png")
result

[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.996078431372549, 0.9882352941176471, 0.996078431372549, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.9921568627450981, 0.984313725490196, 1.0, 1.0, 1.0, 0.9803921568627451, 0.9921568627450981, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.

array([[0., 0., 0., 0., 0., 0., 0., 0., 1., 0.]], dtype=float32)

In [65]:

result=predict_num("C:/Users/Starki/Downloads/zero.png")
result.argmax()

[[[[1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]
   [1.        ]]

  [[1.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [1.        ]]

  [[1.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [0.        ]
   [

0

In [19]:
images={0:"C:/Users/Starki/Downloads/zero.png",
      2:"C:/Users/Starki/Downloads/two.png",
      3:"C:/Users/Starki/Downloads/three.png",
      5:"C:/Users/Starki/Downloads/five.png",
      6:"C:/Users/Starki/Downloads/six.png",
      7:"C:/Users/Starki/Downloads/seven.png",
      8:"C:/Users/Starki/Downloads/eight.png",
      9:"C:/Users/Starki/Downloads/nine.jpg"}
for i in images.keys():
    # #train_generator.class_indices
    result=predict_num(images[i])
    print(i,result.argmax())
    

C:/Users/Starki/Downloads/zero.png
0 0
C:/Users/Starki/Downloads/two.png
2 2
C:/Users/Starki/Downloads/three.png
3 3
C:/Users/Starki/Downloads/five.png
5 5
C:/Users/Starki/Downloads/six.png
6 6
C:/Users/Starki/Downloads/seven.png
7 7
C:/Users/Starki/Downloads/eight.png
8 8
C:/Users/Starki/Downloads/nine.jpg
9 9


D:\codes\ml_and_data_science\mnist\temp\temp.png


3

# final codes in one cell

In [ ]:
import tensorflow as tf
import pandas as pd

train_csv=pd.read_csv("dataset/mnist_train.csv")
test_csv=pd.read_csv("dataset/mnist_test.csv")
print("train data:",train_csv.shape)
print("test data:",test_csv.shape)
print(train_csv.head())
print(test_csv.head())

#splitting data int features and labels
xtrain=train_csv.drop("label",axis=1)
ytrain=train_csv.label
xtest=test_csv.drop("label",axis=1)
ytest=test_csv.label
print("train split:",xtrain.shape,ytrain.shape)
print("test split:",ytest.shape,ytest.shape)

import numpy as np
xtrain=np.array(xtrain).reshape(xtrain.shape[0],28,28,1)
ytrain=np.array(ytrain)
xtest=np.array(xtest).reshape(xtest.shape[0],28,28,1)
ytest=np.array(ytest)
print(xtest.shape)
print(ytest.shape)
print(xtrain.shape)
print(ytrain.shape)

# Making sure that the values are float so that we can get decimal points after division
xtrain = xtrain.astype('float32')
xtest = xtest.astype('float32')
# Normalizing the RGB codes by dividing it to the max RGB value.
xtrain=xtrain/255
xtest=xtest/255

#final tuned model
np.random.seed(42)
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV
from tensorflow.keras.layers import Conv2D,Dropout,Flatten,MaxPooling2D,Dense 
from tensorflow.keras.models import Sequential

model=Sequential()
model.add(Conv2D(filters=64,kernel_size=(3,3),input_shape=(28,28,1)))
#28:number of filters
#kernel_size or filter matrix size
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Flatten())
model.add(Dense(units=256,activation="relu",))
model.add(Dropout(0.2))
model.add(Dense(units=10,activation="softmax"))
model.compile(optimizer="adam", 
              loss='sparse_categorical_crossentropy', 
              metrics=['accuracy'])
model.fit(x=xtrain,y=ytrain, epochs=25)

model.evaluate(xtest,ytest)

#saving and reloading the save model

#serialize model to json file and save weights
from tensorflow.keras.models import model_from_json
model_json=model.to_json()
with open("model.json","w") as json_file:
    json_file.write(model_json)
model.save_weights("model.h5")
print("model saved")

#load model
json_file=open('model.json','r')
loaded_model_json=json_file.read()
json_file.close()
loaded_model=model_from_json(loaded_model_json)
loaded_model.load_weights("model.h5")
print("model loaded")

loaded_model.compile(optimizer="adam",loss='sparse_categorical_crossentropy',metrics=['accuracy'])

score=loaded_model.evaluate(xtest,ytest)
print(score)

#function to predict single image
from tensorflow.keras.preprocessing import image

def predict_num(path):
    
    try:
        test_image=image.load_img(path,target_size=(28,28),color_mode="grayscale")
        test_image=image.img_to_array(test_image)
        test_image=np.expand_dims(test_image,axis=0)
        #train_generator.class_indices
        result=loaded_model.predict(test_image)
       
        return result
              
    except:
         print("invalid address")
            
#testing
images={0:"C:/Users/Starki/Downloads/zero.png",
      2:"C:/Users/Starki/Downloads/two.png",
      3:"C:/Users/Starki/Downloads/three.png",
      5:"C:/Users/Starki/Downloads/five.png",
      6:"C:/Users/Starki/Downloads/six.png",
      7:"C:/Users/Starki/Downloads/seven.png",
      8:"C:/Users/Starki/Downloads/eight.png",
      9:"C:/Users/Starki/Downloads/nine.jpg"}
for i in path.keys():
    # #train_generator.class_indices
    result=predict_num(images[i])
    print(i,result.argmax())